packages/docusaurus/src/server/plugins/init.ts

```diff
@@ -44,39 +45,122 @@ export default function initPlugins({
 
   const plugins: InitPlugin[] = pluginConfigs
     .map((pluginItem) => {
-      let pluginModuleImport: string | undefined;
+      // let pluginModuleImport: string | undefined;
       let pluginOptions: PluginOptions = {};
+      // let plugin: PluginModule | undefined;
+      let validateOptions;
+      // let pluginModule: ImportedPluginModule | undefined;
+      let pluginVersion: DocusaurusPluginVersionInformation;
 
       if (!pluginItem) {
         return null;
       }
 
-      if (typeof pluginItem === 'string') {
-        pluginModuleImport = pluginItem;
-      } else if (Array.isArray(pluginItem)) {
-        [pluginModuleImport, pluginOptions = {}] = pluginItem;
+      function getPlugin(): PluginModule | null {
+        if (Array.isArray(pluginItem)) {
+          if (typeof pluginItem[0] === 'string') {
+            const pluginModuleImport = pluginItem[0];
+            const pluginPath = pluginRequire.resolve(pluginModuleImport);
+            const pluginModule = importFresh<ImportedPluginModule>(pluginPath);
+            return pluginModule?.default || pluginModule;
+          } else if (typeof pluginItem[0] === 'function') {
+            return pluginItem[0];
+          }
+        } else if (typeof pluginItem === 'string') {
+          const pluginModuleImport = pluginItem;
+          const pluginPath = pluginRequire.resolve(pluginModuleImport);
+          const pluginModule = importFresh<ImportedPluginModule>(pluginPath);
+          return pluginModule?.default || pluginModule;
+        } else if (typeof pluginItem === 'function') {
+          return pluginItem;
+        }
+
+        return null;
```

Comment:
Is this case actually possible (assuming we have a good validation?)

If it's not possible, then throw an error like `throw new Error("Unexpected")`.

In general it's a good practice to fail-fast in case of unexpected met, like at the end/default of a switch/case or in the "else" branch of conditionals.

Didn't read but looks like a good article to illustrate the concept: https://levelup.gitconnected.com/switch-guard-in-typescript-how-to-make-sure-all-possible-cases-are-handled-19e86fd15797